pkg/meta/base.go

```diff
@@ -2444,3 +2485,104 @@ LOOP:
 	}
 	return eno
 }
+
+func (m *baseMeta) mergeAttr(ctx Context, inode Ino, set uint16, cur, attr *Attr, now time.Time) (*Attr, syscall.Errno) {
+	dirtyAttr := *cur
+	if (set&(SetAttrUID|SetAttrGID)) != 0 && (set&SetAttrMode) != 0 {
+		attr.Mode |= (cur.Mode & 06000)
+	}
+	var changed bool
+	if (cur.Mode&06000) != 0 && (set&(SetAttrUID|SetAttrGID)) != 0 {
+		clearSUGID(ctx, &dirtyAttr, attr)
+		changed = true
+	}
+	if set&SetAttrGID != 0 {
+		if ctx.Uid() != 0 && ctx.Uid() != cur.Uid {
+			return nil, syscall.EPERM
+		}
+		if cur.Gid != attr.Gid {
+			if ctx.CheckPermission() && ctx.Uid() != 0 && !containsGid(ctx, attr.Gid) {
+				return nil, syscall.EPERM
+			}
+			dirtyAttr.Gid = attr.Gid
+			changed = true
+		}
+	}
+	if set&SetAttrUID != 0 && cur.Uid != attr.Uid {
+		if ctx.Uid() != 0 {
+			return nil, syscall.EPERM
+		}
+		dirtyAttr.Uid = attr.Uid
+		changed = true
+	}
+	if set&SetAttrMode != 0 {
+		if ctx.Uid() != 0 && (attr.Mode&02000) != 0 {
+			if ctx.Gid() != cur.Gid {
+				attr.Mode &= 05777
+			}
+		}
+		if attr.Mode != cur.Mode {
+			if ctx.Uid() != 0 && ctx.Uid() != cur.Uid &&
+				(cur.Mode&01777 != attr.Mode&01777 || attr.Mode&02000 > cur.Mode&02000 || attr.Mode&04000 > cur.Mode&04000) {
+				return nil, syscall.EPERM
+			}
+			dirtyAttr.Mode = attr.Mode
+			changed = true
+		}
+	}
+	if set&SetAttrAtimeNow != 0 || (set&SetAttrAtime) != 0 && attr.Atime < 0 {
```

Comment:
when would be `attr.Atime < 0`?

Reply:
Actually, I don't know when it occurs in the real world, but I see the [setattrStr](https://github.com/juicedata/juicefs/blob/6f8c47cc2cf9a43453e4250abde2b1b85c6024af/pkg/vfs/vfs_unix.go#LL114C37-L114C37) and its unit tests assume it will.